extensions/ql-vscode/src/view/results/raw-results-table.tsx

```diff
@@ -19,158 +18,157 @@ import { onNavigation } from "./results";
+        const currentRow = prevSelectedItem?.row;
+        const nextRow = currentRow === undefined ? 0 : currentRow + rowDelta;
+        if (nextRow < 0 || nextRow >= numberOfAlerts) {
+          return prevSelectedItem;
+        }
+        const currentColumn = prevSelectedItem?.column;
+        const nextColumn =
+          currentColumn === undefined ? 0 : currentColumn + columnDelta;
+        // Jump to the location of the new cell
+        const rowData = resultSet.rows[nextRow];
+        if (nextColumn < 0 || nextColumn >= rowData.length) {
+          return prevSelectedItem;
+        }
+        const cellData = rowData[nextColumn];
+        if (cellData != null && typeof cellData === "object") {
+          const location = tryGetResolvableLocation(cellData.url);
+          if (location !== undefined) {
+            jumpToLocation(location, databaseUri);
+          }
         }
+        scroller.scrollIntoViewOnNextUpdate();
+        return { row: nextRow, column: nextColumn };
+      });
+    },
+    [databaseUri, resultSet, scroller],
+  );
+
+  const handleNavigationEvent = React.useCallback(
+    (event: NavigateMsg) => {
+      switch (event.direction) {
+        case NavigationDirection.up: {
+          navigateWithDelta(-1, 0);
+          break;
+        }
+        case NavigationDirection.down: {
+          navigateWithDelta(1, 0);
+          break;
+        }
+        case NavigationDirection.left: {
+          navigateWithDelta(0, -1);
+          break;
+        }
+        case NavigationDirection.right: {
+          navigateWithDelta(0, 1);
+          break;
+        }
+        default:
+          assertNever(event.direction);
       }
-      this.scroller.scrollIntoViewOnNextUpdate();
-      return {
-        ...prevState,
-        selectedItem: { row: nextRow, column: nextColumn },
-      };
-    });
+    },
+    [navigateWithDelta],
+  );
+
+  React.useEffect(() => {
+    onNavigation.addListener(handleNavigationEvent);
+    return () => {
+      onNavigation.removeListener(handleNavigationEvent);
+    };
+  }, [handleNavigationEvent]);
+
+  let dataRows = resultSet.rows;
+  if (dataRows.length === 0) {
+    return emptyQueryResultsMessage();
   }
 
-  componentDidUpdate() {
-    this.scroller.update();
+  let numTruncatedResults = 0;
+  if (dataRows.length > RAW_RESULTS_LIMIT) {
+    numTruncatedResults = dataRows.length - RAW_RESULTS_LIMIT;
+    dataRows = dataRows.slice(0, RAW_RESULTS_LIMIT);
   }
 
-  componentDidMount() {
-    this.scroller.update();
-    onNavigation.addListener(this.handleNavigationEvent);
+  const tableRows = dataRows.map((row: ResultRow, rowIndex: number) => (
+    <RawTableRow
+      key={rowIndex}
+      rowIndex={rowIndex + offset}
+      row={row}
+      databaseUri={databaseUri}
+      selectedColumn={
+        selectedItem?.row === rowIndex ? selectedItem?.column : undefined
+      }
+      onSelected={setSelection}
+      scroller={scroller}
```

Comment:
(If you take my suggestion of using `useRef` for the scroller, the `RawTableRow` should probably also receive a `ref`. I don't think changing this to `scroller.current` would be correct)

Reply:
I don't quite understand the reasoning here. Could you explain a bit more? My thinking is that passing `scroller.current` here should be the thing to do.

My understanding of `useRef` is that when the value changes it does not trigger a redraw. So if the value did change we wouldn't re-run this code. We also don't ever change the value of `scroller` anyway so it doesn't really matter, but let's ignore that.

Let's say we did change the value and it did redraw for some other reason. In that case we want to pass a new value to `RawTableRow` because the `scroller` object itself won't have changed so it won't see that it needs to redraw `RawTableRow` too and not use a cached value. By passing `scroller.current` we would be passing a new object and it would re-render `RawTableRow` correctly.

Or is there another reason that I'm missing?

Reply:
I'm actually not totally sure about this. [The documentation](https://react.dev/reference/react/useRef) mentions "Do not write or read ref.current during rendering." That would imply that we shouldn't read `scroller.current` anywhere, not even when we pass it along. I think one way around this is the following:

1. Pass the `scroller` ref into the `RawTableRow`
2. Make the `RawTableRow` accept `scroller?: RefObject<ScrollIntoViewHelper>;`
3. Instead of using `ref={props.scroller?.current?.ref(isSelected)}`, we should create a wrapper around this so we don't read `props.scroller.current` during rendering. We should be able to do this using `ref={node => props.scroller?.current?.ref(isSelected)(node)}`, but this doesn't work because the functional refs don't interface well with the class-component refs.

I think this would be the ideal situation to avoid the pitfall mentioned on that documentation page, but we can't do that for now. I would suggest just using the current implementation (i.e. setting `scroller={scroller.current}`) and working on fixing this potential bug later.

Reply:
That's interesting. I didn't know you could pass a function to that `ref=` property. I'm not totally sure if that counts as "during rendering" of not.

My understanding of the danger is that it's around changing ref data and because a component is not entirely defined by its props and state. That means we avoid most if not all of the danger because we don't ever change the ref. I think what we have now is probably ok but just ugly. I'm ok with leaving it as it is for now and coming back to fix it later.

Another different approach I thought of could be https://github.com/github/vscode-codeql/commit/f91f7aa9affe2ebe9372526e335bfb666846a320, avoiding using a ref entirely and instead look up the element by id. I'm not 100% sure this is safe either, since I wonder when the DOM is changed and when `useEffect` hooks run.

Reply:
I believe the DOM is already changed when `useEffect` is run. If you want to run something before the DOM is updated, there's another hook [`useLayoutEffect`](https://react.dev/reference/react/useLayoutEffect) that's made specifically for that purpose. I'm happy to go with this solution for now, but removing the ref and making this easier to understand would definitely be my preference.